DoIP improvements all-around

ferdinandjarisch opened the pull request:
This PR contains multiple fixes that:

* Improve logging output of intermediate results and summaries
* Streamline how connection errors are treating in RoutingActivationType/SourceAddress enumeration
* Avoid stalling enumerations due to dangling socket connections when enumerating SourceAddresses

ferdinandjarisch committed:
fix(doip-discover): Increase log-level of findings

ferdinandjarisch committed:
chore(discover-doip): Use proper repr method for error printing

ferdinandjarisch committed:
chore(discover-doip): ConnectionResetError is a child of ConnectionError

ferdinandjarisch committed:
chore(discover-doip): Streamline and improve logging output

ferdinandjarisch committed:
fix(doip): Add option to close open connections without TIME_WAIT states

ferdinandjarisch committed:
chore(doip): Streamline logging output of recv/sent messages

ferdinandjarisch committed:
chore(doip): Add myself as CodeOwner

rumpelsepp commented:
What' the rational of b8c3ce051463bec271bd73840b6a89502d4337a6? This code seems unfortunate to me. Why is writing random data to a connection needed to close it?

ferdinandjarisch commented:
> What' the rational of [b8c3ce0](https://github.com/Fraunhofer-AISEC/gallia/commit/b8c3ce051463bec271bd73840b6a89502d4337a6)? This code seems unfortunate to me. Why is writing random data to a connection needed to close it?

As mentioned in the log message and according to what you wrote yourself, the rationale is to "to help our kernel realize the connection is closed to avoid TIME_WAIT".

Apparently this is due to implementation details of asyncio's open_connection, having a stream reader/writer on both ends that do not directly communicate, iirc.
Writing EOF alone is not sufficient.

In any case, without this change the DoIP discoverer occupies up to 13.000 Ports while enumerating, as they are not properly closed.

ferdinandjarisch committed:
fix(doip): Only check prevData that was actually received

According to the standard the DoIP entity can respond with only part of
what was initially sent. This is true both for 2019 and 2012 versions of
the standard.

ferdinandjarisch committed:
feat(doip): Expose DoIP protocol version and default to 2019

ferdinandjarisch committed:
fix(doip): Correctly implement GenericDoIPHeaderNACK

ferdinandjarisch committed:
feat(doip): Implement VehicleIdentificationRequests

ferdinandjarisch committed:
feat(doip): Implement DoIPEntityRequests

ferdinandjarisch committed:
feat(doip): Do not hard-fail on unimplemented PayloadTypes

ferdinandjarisch committed:
feat(doip-discover): Gather additional data via UDP

ferdinandjarisch commented on src/gallia/commands/discover/doip.py:
```diff
@@ -188,6 +198,56 @@
+    ) -> None:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
+        sock.settimeout(2)
+        sock.bind(("0.0.0.0", 0))
```
No need to give a specific interface here, since we do not know where the packet will be sent to...

ferdinandjarisch commented:
With recent changes this PR now fully surpasses #506

Additional content includes:

* Fixes treatment of DiagnosticMessageACKs to match requirements from Standard
* Exposes DoIP protocol version to target string
* Auto-detects protocol_version during discovery
* Queries additional content from DoIP entity via DoIPEntityStatusRequests

ferdinandjarisch committed:
feat(doip-discover): Add option to delay subsequent TCP SYNs

ferdinandjarisch committed:
feat(doip-discover): Dynamically adapt RA Request scans

It is up to the DoIP entity to decide whether it first checks for
correct SourceAddress or RoutingActivationType (while the standard
actually puts the check for SourceAddress first...).
This commit minimizes the search space by determining which check is
performed first, and accordingly adapting the following scans.
By doing so it also removes code duplication.

ferdinandjarisch committed:
feat(doip): Add missing RoutingActivationResponseCode

ferdinandjarisch commented:
And even more feats:

* `--tcp-connect-delay` flag to slow down enumerations, which can be useful for not-so-powerful DoIP entities
* Dynamic scanning of RoutingActivationType and SourceAddress, depending on what exactly the DoIP entity prefers

Apart from that some minor fixups for previous commits; so once done this needs to be rebased

rumpelsepp commented:
For the record, discussed via matrix. ACK, but these points are open:

1. Use asyncio for special UDP socket; supported since Python 3.11: https://docs.python.org/3/library/asyncio-eventloop.html#working-with-socket-objects-directly

``` python
socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
sock.setblocking(False)
sock.bind(("0.0.0.0", 0))
loop = asyncio.get_running_loop()
data, addr = await loop.sock_recvfrom(sock, …)
await loop.sock_sendto(sock, …)
```

2. Due to a future refactor, avoid defining new class variables.
3. Make the socket option `SO_LINGER` optional. This is a special case for the discovery scanner and might cause unexpected repercussions for other scanners.

ferdinandjarisch committed:
fix(doip-discover): Improve closing of open connections to avoid TIME_WAIT states

ferdinandjarisch committed:
feat(doip-discover): Unpack discovery response to VehicleAnnouncementMessage

ferdinandjarisch committed:
feat(doip-discover): Expose ProtocolVersion as class variable

ferdinandjarisch committed:
feat(doip): Coherently use RESERVERD and ManufacturerSpecific in enums

ferdinandjarisch commented:
All comments should be addressed.

Maybe let it rest for a day to let me double check tomorrow, otherwise it should be good to merge.

ferdinandjarisch commented:
And for the record: this MR fully surpasses and thus closes #506
